src/common/pool_map.c

```diff
@@ -1422,129 +1459,118 @@ add_domains_to_pool_buf(struct pool_map *map, struct pool_buf *map_buf,
 	}
 
+	if (rc == 0 && !updated)
+		return -DER_ALREADY;
+
 	return rc;
 }
 
 int
 gen_pool_buf(struct pool_map *map, struct pool_buf **map_buf_out, int map_version, int ndomains,
-	     int nnodes, int ntargets, const uint32_t *domains, const d_rank_list_t *ranks,
-	     uint32_t dss_tgt_nr)
+	     int nnodes, int ntargets, const uint32_t *domains, uint32_t dss_tgt_nr)
 {
 	struct pool_component	map_comp;
 	struct pool_buf		*map_buf;
-	struct pool_domain	*found_dom;
 	uint32_t		num_comps;
 	uint8_t			new_status;
-	bool			updated;
 	int			i, rc;
-	uint32_t		num_domain_comps;
-
-	updated = false;
+	uint32_t		num_domain_comps = 0;
+	d_rank_list_t		*ordered_ranks;
 
 	/*
 	 * Estimate number of domains for allocating the pool buffer
 	 */
 	rc = d_fd_get_exp_num_domains(ndomains, nnodes, &num_domain_comps);
 	if (rc != 0) {
-		D_ERROR("Invalid domain array, len=%u\n", ndomains);
+		D_ERROR("failed to calculate number of domains, "DF_RC"\n", DP_RC(rc));
 		return rc;
 	}
 	D_ASSERT(num_domain_comps > 0);
 	num_domain_comps--; /* remove the root domain - allocated separately */
 
+	ordered_ranks = d_rank_list_alloc(nnodes);
+	if (ordered_ranks == NULL)
+		return -DER_NOMEM;
+
 	map_buf = pool_buf_alloc(num_domain_comps + nnodes + ntargets);
 	if (map_buf == NULL)
-		D_GOTO(out_map_buf, rc = -DER_NOMEM);
+		D_GOTO(out_ranks, rc = -DER_NOMEM);
 
-	rc = add_domains_to_pool_buf(map, map_buf, map_version, ndomains,
-				     domains);
+	rc = add_domain_tree_to_pool_buf(map, map_buf, map_version, dss_tgt_nr, ndomains,
+					 domains, ordered_ranks);
 	if (rc != 0)
 		D_GOTO(out_map_buf, rc);
 
 	if (map != NULL) {
 		new_status = PO_COMP_ST_NEW;
-		num_comps = pool_map_find_domain(map, PO_COMP_TP_RANK,
-						 PO_COMP_ID_ALL, NULL);
+		num_comps = pool_map_find_target(map, PO_COMP_ID_ALL, NULL);
 	} else {
 		new_status = PO_COMP_ST_UPIN;
 		num_comps = 0;
 	}
 
-	/* fill nodes */
-	for (i = 0; i < nnodes; i++) {
-		if (map) {
-			found_dom = pool_map_find_node_by_rank(map, ranks->rl_ranks[i]);
-			if (found_dom)
-				continue;
-		}
-
-		updated = true;
-		map_comp.co_type = PO_COMP_TP_RANK;
-		map_comp.co_status = new_status;
-		map_comp.co_index = i + num_comps;
-		map_comp.co_padding = 0;
-		map_comp.co_id = ranks->rl_ranks[i];
-		map_comp.co_rank = ranks->rl_ranks[i];
-		map_comp.co_ver = map_version;
-		map_comp.co_in_ver = map_version;
-		map_comp.co_fseq = 1;
-		map_comp.co_flags = PO_COMPF_NONE;
-		map_comp.co_nr = dss_tgt_nr;
-
-		rc = pool_buf_attach(map_buf, &map_comp, 1 /* comp_nr */);
-		if (rc != 0) {
-			D_ERROR("failed to attach to pool buf, "DF_RC"\n",
-				DP_RC(rc));
-			D_GOTO(out_map_buf, rc);
-		}
-	}
-
-	if (!updated)
-		D_GOTO(out_map_buf, rc = -DER_ALREADY);
-
-	if (map != NULL)
-		num_comps = pool_map_find_target(map, PO_COMP_ID_ALL, NULL);
-	else
-		num_comps = 0;
-
 	/* fill targets */
-	for (i = 0; i < nnodes; i++) {
+	for (i = 0; i < ordered_ranks->rl_nr; i++) {
```

Comment:
In practice we'd expect the number pulled from the tree to be identical to nnodes in length. Maybe this is something we should assert on though.

Reply:
As I think about it I'm wondering if there are scenarios where we'd expect to get a list of ranks that are already in the map... In that case we may not fill it up. Maybe it would be better to update rl_nr to the length actually populated.